### ISSUE TITLE
Provide Content Security Policy to prevent embedding of iframes on unauhorized origins

### ISSUE THREAD

**Issue by jonkoops** [keycloak/keycloak#29782]
Keycloak has several iframes that are embedded into pages of the relying party, for security reasons these iframes should only be allowed to be embedded by trusted origins configured by the user. The specific iframes in question are as follows:

- The [Session Status iframe](https://github.com/keycloak/keycloak/blob/80de3a0a71bb262c590f3bd411c4017f3db0a789/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java#L48-L52), used to implement the [OpenID Connect Session Management](https://openid.net/specs/openid-connect-session-1_0.html) spec.
- The [third-party cookies iframes](https://github.com/keycloak/keycloak/blob/80de3a0a71bb262c590f3bd411c4017f3db0a789/services/src/main/java/org/keycloak/protocol/oidc/endpoints/ThirdPartyCookiesIframeEndpoint.java#L42-L54), used to detect if a user is able to use the cookies of the Keycloak server in a third-party context.

For these resources the [`frame-ancestors`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors) Content Security Policy directive should be set so that only valid origins will be able to embed them. This can be determined as follows:

1. The Keycloak JS adapter embeds the iframes with the [`referrerpolicy`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#referrerpolicy) set to [`strict-origin`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy#strict-origin), which will only send the origin in the `Referer` header when the protocol security level stays the same (HTTPS→HTTPS),
2. The [`Referer`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer) header is read by the Keycloak server from the HTTP request for the iframe, and validated against the configured Web Origins of the clients in the realm (see the [existing code](https://github.com/keycloak/keycloak/blob/80de3a0a71bb262c590f3bd411c4017f3db0a789/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LoginStatusIframeEndpoint.java#L54-L71) for the Session Status iframe for reference).
3. If any client is found with a Web Origin that matches the `Referer` header, the `frame-ancestors` directive should be set to the  [host-source](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#host-source) of the referrer, otherwise the directive should be set to [`none`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors#none).

There is a valid reason that we do not currently implement this functionality. For step 2, it is required to iterate over all clients in the realm, as at that point the client id is not know. This can lead to performance issues where a realm contains many clients. As part of this issue a solution must be implemented to mitigate any performance issues that could lead to a denial of service attack.

**Recommended reading**
- [`Referer` header (MDN)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer)
- [`Referer` header: Privacy and security concerns (MDN)](https://developer.mozilla.org/en-US/docs/Web/Security/Referer_header:_privacy_and_security_concerns)
- [`Referrer-Policy` header (MDN)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy)
- [Iframe `referrerpolicy` attribute (MDN)](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#referrerpolicy)
- [Content Security Policy: `frame-ancestors` directive (MDN)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors)

**Comment by oculos** [comment]
It would be nice to have Web origins for SAML clients. We have legacy clients that need iframes, and it's a pain to have all the domains on a single CSP setting.

Others seem to have that issue as well:
https://github.com/keycloak/keycloak/discussions/12115

**Comment by jonkoops** [comment]
> It would be nice to have Web origins for SAML clients. We have legacy clients that need iframes, and it's a pain to have all the domains on a single CSP setting.

Yeah, this could relate to #29782. Feel free to log a feature request for it.

**Comment by bernhof** [comment]
Does this issue also relate to iframe silent logins (prompt=none), e.g. `GET https://keycloak/realms/foo/protocol/openid-connect/auth?client_id=my_client&prompt=none&...` ?

In this case, Keycloak knows which client is requesting the login via the client_id parameter and should be able to respond with an appropriate `frame-ancestors` CSP directive matching the configured Web Origins of the client.

Also described here: #42587

**Comment by bernhof** [comment]
> Does this issue also relate to iframe silent logins (prompt=none), e.g. `GET https://keycloak/realms/foo/protocol/openid-connect/auth?client_id=my_client&prompt=none&...` ?

I relation to this, Keycloak may in some cases be the upstream IdP that has no information about a downstream broker's clients and their origins. So (obviously, maybe) it should be possible to at least disable this type of dynamic configuration of CSP if no standardised alternatives exist to communicate allowed origins to an IdP (?)

**Comment by msdaly200** [comment]
Just to iterate a couple of gaps for the 3rd-party cookie iframes (3p-cookies/step1.html, step2.html) that was mentioned in #51227 (now closed as duplicate of this issue)

1. IframeUtil.returnIframe() is shared by all three iframe endpoints. (login-status-iframe.html, 3p-cookies/step1.html, 3p-cookies/step2.html) The allowAnyFrameAncestor() call only strips frame-ancestors when the realm value is the default 'self'. If an admin sets a custom value (e.g. 'self' https://abc.com/), the directive leaks through to the 3p-cookies pages, breaking silent-check-SSO. The 3p-cookies pages need frame-ancestors stripped unconditionally, but this should be decoupled from the login-status-iframe behaviour, if it wants to have a proper frame-ancestors.
2. Note that 3p-cookies-step1.html and 3p-cookies-step2.html use window.parent.postMessage(..., "*"), broadcasting the cookie support result to any origin. Unlike login-status-iframe.ftl which uses event.source.postMessage(result, event.origin). This may also need to be changed for hardening.
3. If client_id is added as a parameter to the step1/step2 endpoints, it must be carried through the step1→step2 redirect. Currently window.location = "step2.html" drops all query params.
